sqs/src/main/scala/akka/stream/alpakka/sqs/impl/SqsBatchChangeMessageVisibilityFlowStage.scala

```diff
@@ -25,70 +25,78 @@ import scala.concurrent.{Future, Promise}
-        for (entry <- entries.asScala)
-          log.debug(s"Set visibility timeout for message {} to {}", entry.getReceiptHandle, entry.getVisibilityTimeout)
-        inFlight -= entries.size()
-        checkForCompletion()
-      }
+
       private var changeVisibilityCallback: AsyncCallback[ChangeMessageVisibilityBatchRequest] = _
+
       override def preStart(): Unit = {
         super.preStart()
-        changeVisibilityCallback = getAsyncCallback[ChangeMessageVisibilityBatchRequest](handleChangeVisibility)
+        changeVisibilityCallback = getAsyncCallback[ChangeMessageVisibilityBatchRequest] { request =>
+          val entries = request.getEntries
+          for (entry <- entries.asScala)
+            log.debug("Set visibility timeout for message {} to {}", entry.getReceiptHandle, entry.getVisibilityTimeout)
+          inFlight -= entries.size()
+          checkForCompletion()
+        }
       }
-      override def onPush() = {
-        val messagesIt = grab(in)
-        val messages = messagesIt.toList
-        val nrOfMessages = messages.size
+
+      override def onPush(): Unit = {
+        val actions = grab(in).toList
+        val nrOfActions = actions.size
+
         val responsePromise = Promise[List[SqsAckResult]]
-        inFlight += nrOfMessages
+        inFlight += nrOfActions
 
-        sqsClient
-          .changeMessageVisibilityBatchAsync(
-            new ChangeMessageVisibilityBatchRequest(
-              queueUrl,
-              messages.zipWithIndex.map {
-                case (action, index) =>
-                  new ChangeMessageVisibilityBatchRequestEntry()
-                    .withReceiptHandle(action.message.getReceiptHandle)
-                    .withVisibilityTimeout(action.visibilityTimeout)
-                    .withId(index.toString)
-              }.asJava
-            ),
-            new AsyncHandler[ChangeMessageVisibilityBatchRequest, ChangeMessageVisibilityBatchResult]() {
-              override def onError(exception: Exception): Unit = {
-                val batchException = new SqsBatchException(messages.size, exception)
+        val handler =
+          new AsyncHandler[ChangeMessageVisibilityBatchRequest, ChangeMessageVisibilityBatchResult]() {
+            override def onError(exception: Exception): Unit = {
+              val batchException = new SqsBatchException(actions.size, exception)
+              responsePromise.failure(batchException)
+              failureCallback.invoke(batchException)
+            }
+
+            override def onSuccess(request: ChangeMessageVisibilityBatchRequest,
+                                   result: ChangeMessageVisibilityBatchResult): Unit =
+              if (!result.getFailed.isEmpty) {
+                val nrOfFailedMessages = result.getFailed.size()
+                val batchException: SqsBatchException =
+                  new SqsBatchException(
+                    batchSize = nrOfActions,
+                    cause = new Exception(
```

Comment:
This looks wrong, to change the `SqsBatchException` or introduce a sub-class would be better.